// promise-queue-ts/src/ConcurrentPromiseQueue.ts
import { Queue } from 'queue-typescript';

type Runner<T = any> = () => Promise<T>;
type RunnerContext<T = any> = {
    runner: Runner<T>,
    resolve: (value: T | PromiseLike<T>) => void,
    reject: (reason?: any) => void
}

export class ConcurrentPromiseQueue {

    readonly maxConcurrency: number;

    queue: Queue<RunnerContext> = new Queue();
    runningPromises: number = 0;

    constructor(maxConcurrency: number) {
        this.maxConcurrency = maxConcurrency;
    }

    private checkNext() {
        if(this.queue.length===0) return;
        if(this.runningPromises>=this.maxConcurrency) return;

        const context = this.queue.dequeue();
        try {
            this.runningPromises++;
            context.runner().then(res => {
                context.resolve(res);
                this.runningPromises--;
                this.checkNext();
            }).catch(error => {
                context.reject(error);
                this.runningPromises--;
                this.checkNext();
            });
        } catch (error) {
            context.reject(error);
            this.runningPromises--;
        }

        this.checkNext();
    }

    enqueue<T>(runner: Runner<T>): Promise<T> {
        return new Promise<T>((resolve, reject) => {
            this.queue.enqueue({runner, resolve, reject});
            this.checkNext();
        });
    }

}
