// promise-queue-ts@7.0.25/dist/ConcurrentPromiseQueue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConcurrentPromiseQueue = void 0;
const queue_typescript_1 = require("queue-typescript");
class ConcurrentPromiseQueue {
    constructor(maxConcurrency) {
        this.queue = new queue_typescript_1.Queue();
        this.runningPromises = 0;
        this.maxConcurrency = maxConcurrency;
    }
    checkNext() {
        if (this.queue.length === 0)
            return;
        if (this.runningPromises >= this.maxConcurrency)
            return;
        const context = this.queue.dequeue();
        try {
            this.runningPromises++;
            context.runner().then(res => {
                context.resolve(res);
                this.runningPromises--;
                this.checkNext();
            }).catch(error => {
                context.reject(error);
                this.runningPromises--;
                this.checkNext();
            });
        }
        catch (error) {
            context.reject(error);
            this.runningPromises--;
        }
        this.checkNext();
    }
    enqueue(runner) {
        return new Promise((resolve, reject) => {
            this.queue.enqueue({ runner, resolve, reject });
            this.checkNext();
        });
    }
}
exports.ConcurrentPromiseQueue = ConcurrentPromiseQueue;
